urlfc1c29991b_nil_mish_InJava_tgz-pJ8-edu_usc_cs_nsl_indexing_test_TestTrieProtoJ8/src/edu/usc/cs/nsl/counting/CountingUtil.java

```diff
@@ -0,0 +1,494 @@
+		BufferedWriter bw_output = new BufferedWriter(new FileWriter(output_file,true));
+		HashSet<Integer> index_set = new HashSet<Integer>(num_collisions+10,1);
+		BufferedReader br_collision = new BufferedReader(new FileReader(getCollisionTmpFileName(max_string_size)));
+		String the_line = br_collision.readLine();
+		while(the_line != null){			
+			index_set.add(Integer.parseInt(the_line.replace("\n", "").replace("\r","")));			
+			the_line = br_collision.readLine();
+		}
+		br_collision.close();
+		BufferedReader br;
+		HashMap<String,Long> string_map = new HashMap<String,Long>(num_collisions/10);
+		br = new BufferedReader(new FileReader(getTmpFileName(max_string_size)));
+		the_line = br.readLine();
+		int the_count = 0;
+		while(the_line != null){
+			String the_word = the_line.replace("\n","").replace("\r", "");
+			if(the_word.equals("0017d7d14cbaf2c1c94de80ef039")){
+				System.out.println("is read");
+			}
+			int big_index = (int) Math.abs(getHash(the_word)%big_array_size);
+			if(index_set.contains(big_index)){
+				the_count += 1;
+				if(the_word.equals("0017d7d14cbaf2c1c94de80ef039")){
+					System.out.println("and is added to map");
+				}
+				if(string_map.containsKey(the_word)){
+					long value = string_map.get(the_word);
+					string_map.put(the_word, value+1);
+				}else{
+					string_map.put(the_word, 1L);
+				}
+			}
+			the_line = br.readLine();
+		}
+		for(String a_word: string_map.keySet()){
+			bw_output.write(a_word+"\t"+string_map.get(a_word)+"\n");
+		}
+		System.out.println("The count: "+the_count);
+		bw_output.close();		
+	}
+	
+	
+	/**
+	 * Function to process collisions. Assumption is collisions are small.
+	 * @param max_string_size
+	 * @param output_file
+	 * @param num_collisions
+	 * @param start_index
+	 * @param end_index
+	 * @throws IOException
+	 */
+
+
+	private void processCollision(int max_string_size,String output_file,int num_collisions,int start_index,int end_index) throws IOException{
+		if(num_collisions == 0)
+			return;
+		if(start_index >= end_index)
+			return;
+		int cps = COLLISION_PROCESS_SIZE.get(max_string_size).intValue();
+		if(end_index - start_index <= cps){
+			HashMap<String,Integer> string_map = new HashMap<String,Integer>(num_collisions/10);
+			HashSet<Integer> index_set = new HashSet<Integer>(num_collisions+10,1);
+			BufferedReader br_collision = new BufferedReader(new FileReader(getCollisionTmpFileName(max_string_size)));
+			String the_line = br_collision.readLine();
+			int i = 0;
+			while(the_line != null){
+				if(i < start_index){
+					i++;
+					the_line = br_collision.readLine();
+					continue;
+				}if(i >= end_index){
+					break;
+				}
+				index_set.add(Integer.parseInt(the_line.replace("\n","").replace("\r", "")));
+				the_line = br_collision.readLine();
+				i++;
+				//System.out.println("and here 2?");
+			}
+			br_collision.close();
+			
+			BufferedReader br;
+			br = new BufferedReader(new FileReader(getTmpFileName(max_string_size)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 452, 453, 467. The resource is closed at line 469. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).